HW1/P6/P6.py

```diff
@@ -0,0 +1,54 @@
+from pyspark import SparkContext
+import numpy as np
+from itertools import chain
+import random
+
+sc = SparkContext("local[4]", appName="HW1-6 Markov Shakespeare")
+sc.setLogLevel("ERROR")
+
+txt = open("Shakespeare.txt").readlines()
+txt = list(chain.from_iterable([line.split() for line in txt]))
+txt2 = txt[1:] #Shift text by one word to get subsequent pairings
+txt3 = txt[2:] #Shift text by two words to get third words to be appended to pairings
+
+#Build the text pairings model
+
+words1 = sc.parallelize(txt).filter(lambda x: not x.isdigit()).filter(lambda x: x != x.upper()).zipWithIndex().map(lambda x: (x[1],x[0])).partitionBy(20)
+words2 = sc.parallelize(txt2).filter(lambda x: not x.isdigit()).filter(lambda x: x != x.upper()).zipWithIndex().map(lambda x: (x[1],x[0])).partitionBy(20)
+words3 = sc.parallelize(txt3).filter(lambda x: not x.isdigit()).filter(lambda x: x != x.upper()).zipWithIndex().map(lambda x: (x[1],x[0])).partitionBy(20)
+
+wordTuples = words1.join(words2).join(words3)
+wordpairs_withfollowing = wordTuples.values() #Strip off index
+wordpairs_agg = wordpairs_withfollowing.groupByKey().mapValues(list) #Group by word pairing
+wordpairs_aggCounts = wordpairs_agg.mapValues(lambda l: [(w,l.count(w)) for w in set(list(l))]).cache() #Count occurence of each third word
```

Comment:
You could have mapped each triplet to 1, and then reduce the ones with the same triplet to get the number of appearances.